### PR TITLE
allow */subresource in rbac policy rules

### PR DESCRIPTION
**Pull request by deads2k**
xref #29698
xref #38756
xref #49504
xref #38810

Allow `*/subresource` format in RBAC policy rules to support polymorphic subresources like `*/scale` for HPA.

@DirectXMan12 fyi

```release-note
RBAC PolicyRules now allow resource=`*/<subresource>` to cover `any-resource/<subresource>`.   For example, `*/scale` covers `replicationcontroller/scale`.
```

### CODE REVIEW

File: pkg/apis/rbac/helpers.go [hunk 1]
```diff
@@ -55,12 +55,28 @@ func APIGroupMatches(rule *PolicyRule, requestedGroup string) bool {
...
+		if ruleResource == combinedRequestedResource {
+			return true
+		}
+
+		// We can also match a */subresource.
+		// if there isn't a subresource, then continue
+		if len(requestedSubresource) == 0 {
+			continue
+		}
+		ruleTokens := strings.SplitN(ruleResource, "/", 2)
```
Comment (liggitt) [inline pkg/apis/rbac/helpers.go]: no allocations here, this is a hot spot.

```
if len(ruleResource) == len(requestedSubresource) + 2 && strings.HasPrefix(ruleResource, "*/") && strings.HasSuffix(ruleResource, requestedSubresource)
```

**Comment by deads2k** [comment]
allocations removed, readability reduced.

**Comment by ericchiang** [comment]
/lgtm

cc @DirectXMan12 for HPA updates.

**Comment by liggitt** [comment]
needs a release note
ruleCovers() needs updating (`*/scale` covers `foo/scale`)

**Comment by liggitt** [comment]
Update the godoc on PolicyRule.Resources:
* change `ResourceAll represents all resources.` to `'*' represents all resources in the specified apiGroups`
* add `'*/foo' represents the subresource 'foo' for all resources in the specified apiGroups`

**Comment by liggitt** [comment]
cc @kubernetes/sig-auth-api-reviews

**Comment by enj** [comment]
/lgtm

Not a super big fan of the idea, but I can see how it would be useful.

**Comment by ericchiang** [comment]
Allocs removed, release note added, updated godoc (and removed use of "ResourceAll"), updated default RBAC role.

/lgtm

**Comment by deads2k** [comment]
@liggitt good minus regeneration?

**Comment by deads2k** [comment]
regenerated

**Comment by deads2k** [comment]
last remain package is docs.  That was generated.  tagging approved.

**Pull request merged**